src/Symfony/Component/Messenger/Worker.php

```diff
@@ -135,6 +136,13 @@ private function handleMessage(Envelope $envelope, ReceiverInterface $receiver,
         try {
             $envelope = $this->bus->dispatch($envelope->with(new ReceivedStamp($transportName)));
         } catch (\Throwable $throwable) {
+            $rejectFirst = $throwable instanceof RejectRedeliveredMessageException;
```

Comment:
It seems to me that the BIG important thing here is: if a message is redelivered, don't *actually* try to handle it again (because it'll probably timeout again). Instead, *skip* handling and have it be redelivered.

Question then: to fix this bug, is it actually important to reject this message *before* sending the redelivery? If we manage to skip the AMQP-redelivered message from being handled, isn't it ok if we follow the normal messenger-redelivery logic (redeliver the message and *then* reject it)? Or am I missing something?

If rejecting it *first* is *not* actually important, we could simplify this patch considerably.

thx :)

Reply:
Rejecting redelivered messages first is important. As I described in https://github.com/symfony/symfony/issues/32055#issuecomment-529630654 and #34082, the repeating errors can happen inside the listeners or the retry publishing (anywhere inside the catch block). In this case, the message never gets rejected (as the exception happens before) and would be redelivered forever.

Reply:
See also comment two lines below

Reply:
Hmm, ok. To make it feel like less of a one-off solution, wdyt about a RejectMessageImmediatelyExceptionInterface that we actually look for (then the class implements this)?

Reply:
We only want to reject redelivered messages first and not others to prevent potential loss of messages. So the interface would suggest something we don't want to promote. Also the `RejectRedeliveredMessageException` is not specific to amqp (hence the generic name) and we could use it also for doctrine and other transports. But for doctrine it's not that important as it has a timer in https://github.com/symfony/symfony/blob/c5024bdbd08550307321ed67a658c8442f65a536/src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php#L36 that prevents the queue-blocking (but it does not prevent the potential redelivery loop)